apps/rush-lib/src/cli/actions/BaseRushAction.ts

```diff
@@ -11,19 +12,43 @@ import {
 import RushConfiguration from '../../data/RushConfiguration';
 import EventHooksManager from '../logic/EventHooksManager';
 
+export interface IRushCommandLineActionOptions extends ICommandLineActionOptions {
+  /** If true, no locking mechanism will be enforced when this action is run */
+  safeForSimultaneousRushProcesses?: boolean;
+}
+
 /**
  * The base Rush action that all Rush actions should extend.
  */
 export abstract class BaseRushAction extends CommandLineAction {
   private _rushConfiguration: RushConfiguration;
   private _eventHooksManager: EventHooksManager;
+  private _safeForSimultaneousRushProcesses: boolean;
 
-  constructor(options: ICommandLineActionOptions) {
+  constructor(options: IRushCommandLineActionOptions) {
     super(options);
+    this._safeForSimultaneousRushProcesses = !!options.safeForSimultaneousRushProcesses;
   }
 
   protected onExecute(): void {
     this._ensureEnvironment();
+
+    if (!this._safeForSimultaneousRushProcesses) {
+      const lockFilePath: string = path.join(
+        this.rushConfiguration.commonTempFolder,
+        'rush.lock'
+      );
+      if (fsx.existsSync(lockFilePath)) {
+        fsx.removeSync(lockFilePath);
+      }
+      try {
+        fsx.openSync(lockFilePath, 'wx');
+      } catch (error) {
+        console.log(`Another rush command is already running in this repository!`);
+        process.exit(1);
```

Comment:
>process.exit(1); [](start = 8, length = 16)

We should work to eliminate `process.exit()` from the Rush code base.  It's ungraceful.  The right approach is to assign process.exitCode and return from the entry point.

Reply:
We should do that in a different PR

---
In reply to: [165497011](https://github.com/Microsoft/web-build-tools/pull/517#discussion_r165497011) [](ancestors = 165497011)